sdk/core/azure-core/src/logger.cpp

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/core/logger.hpp"
+#include "azure/core/internal/log.hpp"
+
+#include <mutex>
+#include <shared_mutex>
+
+using namespace Azure::Core;
+using namespace Azure::Core::Internal;
+
+namespace {
+static std::shared_timed_mutex g_logListenerMutex;
+static Logger::Listener g_logListener(nullptr);
+} // namespace
+
+std::atomic<bool> Log::g_isLoggingEnabled(false);
+std::atomic<Log::LogLevelInt> Log::g_logLevel(static_cast<LogLevelInt>(Logger::Level::Warning));
+
+inline void Log::ToggleLogging(bool isEnabled) { g_isLoggingEnabled = isEnabled; }
```

Comment:
If we keep this (and I need to undersatnd the usage of it), then it should be renamed sisnce it doesn't actually Toggle; it sets, so EnableLogging(bool isEnabled)

Reply:
I agree with the naming, it is probably better (I was thinking of it as "toggling on" and "toggling off", but "enable on"/"enable off" is less confusing).

This is needed, because it sets an atomic boolean variable, which is lock free.
The moment when our code tries to determine whether it should log:
1. It "calls" an **inline** function `ShouldWrite(Logger::Level logLevel)`
2. `ShouldLog(Level)` "calls" an **inline** function `GetLogListener(Logger::Level logLevel)`, and compares it to nullptr.
3. `GetLogListener(Logger::Level logLevel)` checks an atomic and lock free boolean static variable, `g_isLoggingEnabled`. If it is false, returns nullptr.
4. If `g_isLoggingEnabled` is true, it checks an atomic and lock free variable `g_logLevel`: logLevel of the message that we are trying to log should be `<= g_logLevel`.

As you can see, `ShouldWrite()` check is lock-free. Plus, it gets inlined, so the code ends up being as simple as checking a boolean variable, which is as fast as you can get. Maximum, you have two checks, when `g_isLoggingEnabled == true`. A second check is integer `=<` comparison, and is also inlined.

`g_isLoggingEnabled` gets set by `Logger::SetListener(listener)` : if (listener) is nullptr, `g_isLoggingEnabled` gets set to false, otherwise it is true. `g_isLoggingEnabled` can be read as "is g_logListener a non-null pointer"? Only that it is lock-free (we can't check `g_logListener` for being null without a lock. Even though it is an inexpensive lock, especially now, because we now use shared lock when reading, it is still not as fast as lock-free inline comparison (probably something like `CMP g_isLoggingEnabled, 0; JE :should_not_write; CMP EAX, g_logLevel; JG :should_not_write`). @ahsonkhan was suggesting that in WinHttp, logging may get responsible for 80% slowdown due to mutex.

Reply:
> was suggesting that in WinHttp, logging may get responsible for 80% slowdown due to mutex.

Just to clarify, I think @vhvb1989 mentioned something about the perf framework showing a regression when we use `Context` (see https://github.com/Azure/azure-sdk-for-cpp/issues/1589). I didn't measure with winhttp yet (it doesn't log anything atm anyway), so I am not sure where that 80% came from.